src/rest-devices.c

```diff
@@ -0,0 +1,575 @@
+            }
+
+            if (base64_decode(string, binary_buffer, &length))
+            {
+                goto abort;
+            }
+
+            entry->psk = (uint8_t *)malloc(length);
+            if (entry->psk == NULL)
+            {
+                goto abort;
+            }
+
+            memcpy(entry->psk, binary_buffer, length);
+            entry->psk_len = length;
+            count++;
+        }
+        if ((key = json_object_get(value, "psk_id")) != NULL)
+        {
+            string = json_string_value(key);
+            if (string == NULL)
+            {
+                goto abort;
+            }
+
+            if (base64_decode(string, NULL, &length))
+            {
+                goto abort;
+            }
+
+            if (base64_decode(string, binary_buffer, &length))
+            {
+                goto abort;
+            }
+
+            entry->psk_id = (uint8_t *)malloc(length);
+            if (entry->psk_id == NULL)
+            {
+                goto abort;
+            }
+
+            memcpy(entry->psk_id, binary_buffer, length);
+            entry->psk_id_len = length;
+            count++;
+        }
+
+abort:
+        if (count != 3)
+        {
+            rest_devices_delete_list(extension, entry);
+            return -1;
+        }
+        rest_list_add(extension, entry);
+    }
+
+    rest_list_extend(list, extension);
+    return 0;
+}
+
+static int rest_devices_update_list(rest_list_t *list, json_t *jdevice)
+{
+    const char *string;
+    json_t *jstring;
+    rest_list_entry_t *device_entry;
+    database_entry_t *device_data;
+    uint8_t *oldpsk, *oldid;
+    uint8_t binary_buffer[512];
+    size_t length;
+
+    jstring = json_object_get(jdevice, "uuid");
+    string = json_string_value(jstring);
+
+    for (device_entry = list->head; device_entry != NULL; device_entry = device_entry->next)
+    {
+        device_data = (database_entry_t *)device_entry->data;
+
+        if (strcmp(device_data->uuid, string) == 0)
+        {
+            jstring = json_object_get(jdevice, "psk");
+            string = json_string_value(jstring);
+
+            base64_decode(string, NULL, &length);
+            if (base64_decode(string, binary_buffer, &length))
+            {
+                return -1;
+            }
+
+            oldpsk = device_data->psk;
+            device_data->psk = malloc(length);
+            if (device_data->psk == NULL)
+            {
+                device_data->psk = oldpsk;
+                return -1;
+            }
+            memcpy(device_data->psk, binary_buffer, length);
+
+            jstring = json_object_get(jdevice, "psk_id");
+            string = json_string_value(jstring);
+
+            base64_decode(string, NULL, &length);
+            if (base64_decode(string, binary_buffer, &length))
+            {
+                return -1;
+            }
+
+            oldid = device_data->psk_id;
+            device_data->psk_id = malloc(length);
```

Comment:
When allocating data specify not only length, but type too, just in case something changes in implememtation:

`malloc(sizeof(data_type_t) * length)`

Reply:
sizeof is redundant in byte buffer allocations.

Reply:
understood